### PR TITLE
Fix setProps when IDs aren't provided

### PR DESCRIPTION
**Pull request by chriddyp**
Fixes #165

### CODE REVIEW

File: src/TreeContainer.js [hunk 1]
```diff
@@ -87,14 +90,16 @@ CheckedComponent.propTypes = {
...
             return null;
         }
 
         return Array.isArray(components) ?
-            map(createContainer, components) :
-            createContainer(components);
+            addIndex(map)(
+                (component, i) => createContainer(component, concat(path, ['props', 'children', i])),
+                components
+            ) : createContainer(components, concat(path, ['props', 'children']));
```
Comment (alexcjohnson) [inline src/TreeContainer.js]: Non-🚫 but to my eye ramda makes this much harder to read, vs vanilla
```
components.map((component, i) =>
    createContainer(component, concat(path, ['props', 'children', i]))
)
```

File: tests/test_render.py [hunk 2]
```diff
@@ -2960,3 +2961,27 @@ def display_content(pathname):
...
+            app,
+            debug=True,
+            use_reloader=False,
+            use_debugger=True,
+            dev_tools_hot_reload=False,
+        )
+
+        self.wait_for_element_by_css_selector('#id').send_keys('hello input with ID')
+
+        self.wait_for_element_by_css_selector('#container input').send_keys('hello input without ID')
```
Comment (alexcjohnson) [inline tests/test_render.py]: Just because of the issues we've had with percy autoapproval, would you mind testing the desired outcome in code here?

Comment (chriddyp) [reply tests/test_render.py]: Yeah good idea - easy enough to do here anyway.

**Comment by alexcjohnson** [comment]
@chriddyp this is awesome (my :dancer: stands BTW, despite the extra discussion). But @byronz merged some of the `if(id)` hacks in https://github.com/plotly/dash-core-components/pull/523, do we need to remove them now? Seems like they could actually be harmful once this PR is merged.

**Comment by chriddyp** [comment]
Yeah - I think I'll follow up with a PR that cleans up some of the DCC logic. I think that the `id => setState` is actually still "safe" but it could be cleaned up - many of the components can become functional & stateless

**Pull request merged**